Bio/SeqIO/AceIO.py

```diff
@@ -15,49 +15,68 @@
-
-    >>> from Bio import SeqIO
-    >>> with open("Ace/contig1.ace") as handle:
-    ...     for record in SeqIO.parse(handle, "ace"):
-    ...         print("%s ...%s..." % (record.id, record.seq[85:95]))
-    ...         print(record.letter_annotations["phred_quality"][85:95])
-    ...         print(max(record.letter_annotations["phred_quality"]))
-    Contig1 ...AGAGG-ATGC...
-    [57, 57, 54, 57, 57, 0, 57, 72, 72, 72]
-    90
-    Contig2 ...GAATTACTAT...
-    [68, 68, 68, 68, 68, 68, 68, 68, 68, 68]
-    90
-
-    """
-    for ace_contig in Ace.parse(source):
+from .Interfaces import _TextIOSource
+from .Interfaces import SequenceIterator
+
+
+class AceIterator(SequenceIterator):
+    """Return SeqRecord objects from an ACE file."""
+
+    def __init__(
+        self,
+        source: _TextIOSource,
+    ) -> None:
+        """Iterate over SeqRecord objects read from an ACE file.
+
+        Arguments:
+         - source - input stream opened in text mode, or a path to a file
+
+        This uses the Bio.Sequencing.Ace module to do the hard work.  Note that
+        by iterating over the file in a single pass, we are forced to ignore any
+        WA, CT, RT or WR footer tags.
+
+        Ace files include the base quality for each position, which are taken to
+        be PHRED style scores. Just as if you had read in a FASTQ or QUAL file
+        using PHRED scores using Bio.SeqIO, these are stored in the SeqRecord's
+        letter_annotations dictionary under the "phred_quality" key.
+
+        >>> from Bio import SeqIO
+        >>> with open("Ace/consed_sample.ace") as handle:
+        ...     for record in SeqIO.parse(handle, "ace"):
+        ...         print("%s %s... %i" % (record.id, record.seq[:10], len(record)))
+        ...         print(max(record.letter_annotations["phred_quality"]))
+        Contig1 agccccgggc... 1475
+        90
+
+        However, ACE files do not include a base quality for any gaps in the
+        consensus sequence, and these are represented in Biopython with quality
+        of zero. Using zero is perhaps misleading as there may be very strong
+        evidence to support the gap in the consensus. Previous versions of
+        Biopython therefore used None instead, but this complicated usage,
+        and prevented output of the gapped sequence as FASTQ format.
+
+        >>> from Bio import SeqIO
+        >>> with open("Ace/contig1.ace") as handle:
+        ...     for record in SeqIO.parse(handle, "ace"):
+        ...         print("%s ...%s..." % (record.id, record.seq[85:95]))
+        ...         print(record.letter_annotations["phred_quality"][85:95])
+        ...         print(max(record.letter_annotations["phred_quality"]))
+        Contig1 ...AGAGG-ATGC...
+        [57, 57, 54, 57, 57, 0, 57, 72, 72, 72]
+        90
+        Contig2 ...GAATTACTAT...
+        [68, 68, 68, 68, 68, 68, 68, 68, 68, 68]
+        90
+
+        """
+        super().__init__(source, mode="t", fmt="ACE")
```

Comment:
SeqIO is using lower case format names, so "ace".

Reply:
Are you sure? The others use a mixture of upper and lower case:
```
$ grep super *.py | grep fmt
AbiIO.py:        super().__init__(source, mode="b", fmt="ABI")
AceIO.py:        super().__init__(source, mode="t", fmt="ACE")
FastaIO.py:        super().__init__(source, mode="t", fmt="Fasta")
FastaIO.py:        super().__init__(source, mode="t", fmt="FASTA")
FastaIO.py:        super().__init__(source, mode="t", fmt="FASTA")
FastaIO.py:        super().__init__(source, mode="t", fmt="Fasta")
GckIO.py:        super().__init__(source, mode="b", fmt="GCK")
IgIO.py:        super().__init__(source, mode="t", fmt="IntelliGenetics")
InsdcIO.py:        super().__init__(source, mode="t", fmt="GenBank")
InsdcIO.py:        super().__init__(source, mode="t", fmt="EMBL")
InsdcIO.py:        super().__init__(source, mode="t", fmt="IMGT")
InsdcIO.py:        super().__init__(source, mode="t", fmt="GenBank")
InsdcIO.py:        super().__init__(source, mode="t", fmt="EMBL")
NibIO.py:        super().__init__(source, mode="b", fmt="Nib")
PdbIO.py:        super().__init__(source, mode="t", fmt="PDB")
PirIO.py:        super().__init__(source, mode="t", fmt="Pir")
QualityIO.py:        super().__init__(source, mode="t", fmt="Fastq")
QualityIO.py:        super().__init__(source, mode="t", fmt="QUAL")
SeqXmlIO.py:        super().__init__(stream_or_path, mode="b", fmt="SeqXML")
SffIO.py:        super().__init__(source, mode="b", fmt="SFF")
SnapGeneIO.py:        super().__init__(source, mode="b", fmt="SnapGene")
TabIO.py:        super().__init__(source, mode="t", fmt="Tab-separated plain-text")
TwoBitIO.py:        super().__init__(source, mode="b", fmt="twoBit")
XdnaIO.py:        super().__init__(source, mode="b", fmt="Xdna")
```

Reply:
Ideally`fmt` would be an abstract class variable instead of an instance property.
Anyway, something for later.